### PR TITLE
Go lang std template initial support

### PR DESCRIPTION
**Pull request by jhsx**
Hi everyone, this pull request includes an initial support for the text/template package.

### CODE REVIEW

File: .idea/codeStyleSettings.xml [hunk 1]
```diff
@@ -133,7 +133,6 @@
           <option name="ALIGN_MULTILINE_THROWS_LIST" value="true" />
           <option name="ALIGN_MULTILINE_EXTENDS_LIST" value="true" />
           <option name="ALIGN_MULTILINE_PARENTHESIZED_EXPRESSION" value="true" />
-          <option name="ALIGN_SUBSEQUENT_SIMPLE_METHODS" value="true" />
```
Comment (dlsniper) [inline .idea/codeStyleSettings.xml]: Please revert this.

File: .idea/vcs.xml [hunk 2]
```diff
@@ -12,5 +12,6 @@
   </component>
   <component name="VcsDirectoryMappings">
     <mapping directory="$PROJECT_DIR$" vcs="Git" />
+    <mapping directory="$PROJECT_DIR$" vcs="Git" />
```
Comment (dlsniper) [inline .idea/vcs.xml]: Please revert this

**Comment by dlsniper** [comment]
Please remove: JFlex.jar from the PR, it's not needed.

File: resources/META-INF/plugin.xml [hunk 3]
```diff
@@ -93,8 +93,17 @@
     <psi.fileReferenceHelper implementation="com.goide.psi.impl.imports.GoImportReferenceHelper"/>
     <fileContextProvider implementation="com.goide.psi.impl.imports.GoFileContextProvider"/>
 
-    <lang.parserDefinition language="go" implementationClass="com.goide.GoParserDefinition"/>
+
+    <lang.fileViewProviderFactory language="tpl" implementationClass="com.goide.gotemplate.file.GoTemplateFileViewProviderFactory" />
```
Comment (dlsniper) [inline resources/META-INF/plugin.xml]: I'm not sure if the language name is tpl, maybe gotpl to be even more specific?

File: src/com/goide/GoFileType.java [hunk 4]
```diff
@@ -62,8 +62,4 @@ public String getCharset(@NotNull VirtualFile file, @NotNull byte[] content) {
     return CharsetToolkit.UTF8;
   }
 
-  @Override
-  public Charset extractCharsetFromFileContent(@Nullable Project project, @Nullable VirtualFile file, @NotNull CharSequence content) {
```
Comment (dlsniper) [inline src/com/goide/GoFileType.java]: Why remove this from the Go files? They should always be UTF-8 per language specs (or do I miss something?)

**Comment by dlsniper** [comment]
One final comment from my understanding of how the project is structured would be to convert this to a gradle module, like `google-app-engine` for example. @stuartcarnie might be able to help you out on this one.

Otherwise, thank you for this, it's really nice to have it! I'll let @ignatov and @zolotov properly review this!

Another thing to have in mind would be the `html/package` package which is basically `text/template` but for html usage.

**Comment by jhsx** [comment]
@dlsniper Sorry for the mess, i will clean up the tree and push it again, some of those files i don't remember to editing they.
the Pr is not yet in the state for merging.

I do have few issues for discussion that i will list following.
The language name as you mentioned in your comment.
Default File extension and Icon.
Syntax Coloring in a different PR.
Should i create a another module for this ? but the generated source will be shared?

File: idea-flex.skeleton [hunk 5]
```diff
@@ -0,0 +1,251 @@
+  /** initial size of the lookahead buffer */
```
Comment (ignatov) [inline idea-flex.skeleton]: There's no need to have a skeleton file, please see `gen_lexer.sh`.

File: src/com/goide/gotemplate/GoTemplateLanguage.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import com.intellij.openapi.fileTypes.FileType;
+import com.intellij.openapi.fileTypes.LanguageFileType;
+import com.intellij.openapi.fileTypes.StdFileTypes;
+import com.intellij.psi.templateLanguages.TemplateLanguage;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Created by josesantos on 21/12/15.
+ */
```
Comment (ignatov) [inline src/com/goide/gotemplate/GoTemplateLanguage.java]: Please remove such comment with author according the code style.

File: src/com/goide/lexer/_Go_templateLexer.java~ [hunk 7]
```diff
@@ -0,0 +1,728 @@
+/* The following code was generated by JFlex 1.4.3 on 24-12-2015 16:47 */
```
Comment (ignatov) [inline src/com/goide/lexer/_Go_templateLexer.java~]: Do we actually need this?

File: src/com/goide/gotemplate/parser/GoTemplateParserUtil.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package com.goide.gotemplate.parser;
+
+import com.intellij.lang.parser.GeneratedParserUtilBase;
+
+/**
+ * Created by josesantos on 21/12/15.
+ */
+public class GoTemplateParserUtil extends GeneratedParserUtilBase {
```
Comment (ignatov) [inline src/com/goide/gotemplate/parser/GoTemplateParserUtil.java]: It's possible to use a basic `com.intellij.lang.parser.GeneratedParserUtilBase` as a parser util class in the bnf file.

**Comment by ignatov** [comment]
> but the generated source will be shared?
Yes, it should be shared.

**Comment by jhsx** [comment]
Sorry close by mistake

**Pull request closed without merging**